torch/_functorch/aot_autograd.py

```diff
@@ -908,34 +958,89 @@ def joint_forward_backward(
+        # We need to ensure that the inputs we pass to autograd.grad() are the *original*
+        # inputs, and not their mutated values.
+        if maybe_tangents is not None:
+            primals_after_cloning = [
+                maybe_to_fresh_input(i, t) for i, t in enumerate(primals)
+            ]
         else:
-            return t
-
-    def from_fun(t):
-        if not isinstance(t, Tensor) or not torch._is_functional_tensor(t):
-            return t
-        torch._sync(t)
-        return torch._from_functional_tensor(t)
-
-    def functionalized_joint(
-        primals: List[Any], tangents: List[Any]
-    ) -> Tuple[List[Any], List[Any]]:
+            primals_after_cloning = primals
+        outs = flat_fn_with_synthetic_bases_expanded(primals, primals_after_cloning, maybe_tangents)
+        return outs
 
+    def functionalized_f_helper(primals_and_maybe_tangents):
+        if trace_joint:
+            primals, maybe_tangents = primals_and_maybe_tangents
+        else:
+            primals = primals_and_maybe_tangents
+            maybe_tangents = None
         # Wrap inputs into functional wrappers
-        f_primals, f_tangents = pytree.tree_map(to_fun, (primals, tangents))
+        f_primals = pytree.tree_map(to_fun, primals)
+        f_tangents = None if maybe_tangents is None else pytree.tree_map(to_fun, maybe_tangents)
         torch._enable_functionalization(reapply_views=True)
         try:
             # Run the joint
-            f_outs = joint_forward_backward(f_primals, f_tangents)
+            f_outs = flat_fn_no_input_mutations(f_primals, f_tangents)
         finally:
             torch._disable_functionalization()
 
+        if keep_input_mutations:
+            # Note: This is a bit annoying. There's a layering issue here, where:
+            # (1) functionalization needs to operate on **synthetic base** inputs, before unpacking them into the "real" inputs.
+            # (2) For keep_input_mutations, we support tracing a call to copy_() directly on mutated inputs.
+            #     However, we **only** want to support this for inputs that have data-only (and no metadata) mutations,
+            #     because inductor (and backends in generally) would prefer not to see these (e.g. as_strided_(), resize_()).
+            #     This makes it pretty difficult for this logic to operate on synthetic bases.
+            #     For example, suppose "x" and "y" are aliased inputs, and we witness:
+            #       x.add_(1)
+            #       y.t_()
+            #     We don't actually want the compiler to witness a copy_() on the synthetic base here: it should
+            #     just see the copy_() directly on x!
+            #     For now, to do this, I'm manually unpacking the synthetic bases again in this function.
+            primals_unpacked = unpack_synthetic_bases(primals, synthetic_base_info)
+            f_primals_unpacked = unpack_synthetic_bases(f_primals, synthetic_base_info)
```

Comment:
See this comment for details (cc @wconstab since this is related to what you referenced in the meeting).

If you try to compile a kernel like this:
```
def f(x, y):
    x.mul_(2)
    y.mul_(3)
    return x, y

a = torch.ones(1'000'000)
x, y = out(a[0:9], a[1:10])
```

Then:
(1) AOTAutograd will do all of the fancy logic of creating a graph that takes a single argument, corresponding to a "synthetic base" of the aliased x and y arguments
(2) The logic here is what causes us to emit two `copy_()`'s into the graph, for each tiny slice, instead of one massive copy for the (in this case much larger) base.

I think that there are arguments in either direction: in this example, putting the copy_() nodes on the smaller views is more performant, but it means inductor needs to worry about the ordering of the copy_() calls. I'm mainly doing this because it is what AOTAutograd already does today, just in the opaque epilogue.

We could switch things up and make the copies happen on the larger tensor, it would just require some more work - we'd need to actually generate  the "updated synthetic base" that we copy from, which doesn't currently happen.

Reply:
i am not sure its worth doing that. i just wanted to clarify the behavior.  maybe it is not hard for inductor to deal with the copy calls.